src/content/posts/2021/pitfalls-of-unoptimzied-nuget-packages/index.md

```diff
@@ -0,0 +1,177 @@
+---
+title: "Pitfalls of Unoptimized NuGet Packages"
+description: "Avoid using unoptimized .NET assemblies with UnoptimizedAssemblyDetector."
+date: 2021-05-21T16:20:00-04:00
+tags: [".NET", "Assembly", "NuGet", "Optimization"]
+author: "Bruno Garcia"
+showToc: false
+TocOpen: false
+draft: false
+hidemeta: true
+comments: false
+type: posts
+
+cover:
+    image: "snail-cover.jpg"
+    relative: true
+    alt: "A Snail."
+
+resources:
+- src: 'davidfowl-nuget-package-debug-dark.png'
+- src: 'davidfowl-nuget-package-debug-light.png'
+- src: 'dnx-release-build-pack-dark.png'
+- src: 'dnx-release-build-pack-light.png'
+- src: 'unoptimized-assembly-detected.gif'
+
+---
+
+## tl;dr
+
+You might unknowingly be using an unoptimized dependency which could impact the performance of your app.
+Now, it's easy to get a warning if that happens, at build time,
+using [the NuGet package `UnoptimizedAssemblyDetector`](https://www.nuget.org/packages/UnoptimizedAssemblyDetector/):
+
+```xml
+<PackageReference Include="UnoptimizedAssemblyDetector" Version="0.0.5">
+  <PrivateAssets>all</PrivateAssets>
+</PackageReference>
+```
+
+After you add this NuGet package, a warning will be included in your build if any unoptimized assembly is detected.
+
+[![unoptimized-assembly-detected](unoptimized-assembly-detected.gif)](unoptimized-assembly-detected.gif)
+
+The source code of project is [hosted on GitHub](https://github.com/bruno-garcia/unoptimized-assembly-detector) ⭐.
+
+## What's the pitfall?
+
+The default build configuration for `dotnet pack` and `dotnet publish` is not `Release` as you might assume or expect.
+When publishing NuGet packages to _nuget.org_, one must explicitly make it a release package, like: `dotnet pack -c Release`.
+In this blog post you'll learn that's not always the case. Even for those who know, accidents happens and a misconfigured build script can result in a `Debug` build ending up on _nuget.org_.
+
+The default value being `Debug` isn't alone the reason why this is a pitfall.
+There's no warning of any kind from the point you create the NuGet package, through uploading to _nuget.org_ all the way to restoring it for your project.
+
+## Why should I care?
+
+If you need convincing, keep reading and I'll do my best.
+On the other hand if you're already convinced, feel free to skip ahead to [Sensible defaults](#sensible-defaults).
+If you don't care if the code you run was compiled with optimizations, it's unlikely this blog post is for you.
+
+### Optimizations
```

Comment:
This should come right after intro in my opinion. This would serve as a good section to explain how does Debug build end up producing an unoptimzed build (i.e. what compiler options does it pass). Ideally, it would be nice to compare IL code from the same optimized and unoptimized program. I would focus this section on "what does being optimized mean and how does this get handled by the compiler" and move the performance aspects to a different section.

Reply:
Agree it would be nice but I don't feel like putting the effort. I'm sure someone can write a nice blog post about this topic alone.